torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
 
         else:
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
+def first_one(n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        return [Equality(arg_type.__args__[0], n.type.__args__[0])]
```

Comment:
So this is just saying that the batch size at input needs to match the batch size at output?

Can we encode e.g. that the channel dimensions at input/output need to match the parameter value? Or is that not necessary?

Reply:
We already have this in the initial pass

```
        w_in = arg_type.__args__[3]
        h_in = arg_type.__args__[2]
        h_out = calculate_out_dimension(h_in, module_instance, 0)
        w_out = calculate_out_dimension(w_in, module_instance, 1)
        new_type = TensorType((arg_type.__args__[0], module_instance.out_channels, h_out, w_out))
```

So I think we already have that part but maybe we consider moving it here at some point so we can do one pass instead of two passes. I haven't worked out what all the constraints would look like in that case though, so I'll think about it.